### PR TITLE
Add MoE matmul implementation

### PR DESCRIPTION
**Pull request by RissyRan**
# Description
- Add matmul implementation to **eventually replace** for loop in MoE models (as a alternative milestone for Megablox due to the blocker summarized [here](https://docs.google.com/document/d/1W5rNemv5D8oqAG0nHIhX8y8mq1VKKGuXDjKV5a6_wX0/edit?tab=t.0#heading=h.xgjl2srtytjt)). The ultimate goal is to apply Megablox implementation.
- Add a flag `moe_matmul` for now instead of fully replacement, with further optimization (in future PRs).

# Test

At high level:
- Add forward_check test for loop implementation ([logit value check](https://paste.googleplex.com/5912253899735040#l=34), [numbers on TPU vs. CPU](https://docs.google.com/spreadsheets/d/1o2RhkRpPXL5WD7FYb9kgm64qjuifOt0E0nfe3dc0YZE/edit?resourcekey=0-HGJ3wSqnzfMQ52O_Pf10sQ#gid=1640920960)]
- unit test `moe_test.py` to verify output matches between for loop and matmul
- end-to-end test for small model size: token matches between for loop and matmul, logits diverge widely after layer 18
- end-to-end test for original model size as below:

Test links:
* end-to-end for loop (current implementation): [test](https://pantheon.corp.google.com/kubernetes/service/us-central2/v4-bodaborg/default/ran-matmul-decode-matmul-5/logs?e=13803378&mods=allow_workbench_image_override&project=tpu-prod-env-multipod)

```
Input `[INST] I love to [/INST]` -> `That's great to hear! I love to learn new things`
```
* end-to-end matmul (my change): [test](https://pantheon.corp.google.com/kubernetes/service/us-central2/v4-bodaborg/default/ran-matmul-decode-matmul-7/logs?e=13803378&mods=allow_workbench_image_override&project=tpu-prod-env-multipod)

```
Input `[INST] I love to [/INST]` -> `That's great to hear! Could you please tell me what`
```

### CODE REVIEW

**Review by gobbleturk — CHANGES_REQUESTED** [review]

File: MaxText/tests/forward_pass_logit_checker.py [hunk 1]
```diff
@@ -107,8 +107,9 @@ def main(config):
     )
 
     max_logging.log(f"{golden_logits[0] =}, {full_train_logits[0, 0, :]=}")
+    # compare first 4 token logits
     assert jax.numpy.allclose(
-            full_train_logits[0, :, :], golden_logits, rtol=1e-01, atol=1e-01, equal_nan=False
+            full_train_logits[0, :4, :], golden_logits[:4, :], rtol=float(config.rtol), atol=float(config.atol), equal_nan=False
```
Comment (A9isha) [inline MaxText/tests/forward_pass_logit_checker.py:112]: we shouldn't be hardcoding the number of tokens to compare, it should be coming from the number of tokens in the golden data

**Review by A9isha — CHANGES_REQUESTED** [review]

File: MaxText/tests/forward_pass_logit_checker.py [hunk 2]
```diff
@@ -75,7 +75,7 @@ def get_data(golden_data, golden_data_index, config):
 def main(config):
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply MaxText/tests/forward_pass_logit_checker.py:75]: Could you explain a little bit how we distinguish those? 

I was trying, and hit the issue:

```
parser.add_argument("--atol", type=float, required=False, default=0.1)
args, unknown = parser.parse_known_args()
```

File: MaxText/tests/moe_test.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+import common_types
+import flax.linen as nn
+
+
+Array = common_types.Array
+Config = common_types.Config
+DType = common_types.DType
+NdInitializer = initializers.NdInitializer
+
+class MoeLoopBlock(nn.Module):
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply MaxText/tests/moe_test.py:36]: Yeah, those logic are same from existing [logic](https://github.com/google/maxtext/blob/main/MaxText/layers/mistral.py#L136-L165), but move into a class called `MoeLoopBlock`. Actually, this unit test may not be needed in the repo. I agree, let's test with logits with each implementation directly.

File: MaxText/tests/forward_pass_logit_checker.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (A9isha) [reply MaxText/tests/forward_pass_logit_checker.py:75]: one naive way is to not pass the whole `sys.argv` to `pyconfig.initialize` if there's a convention of atol/rtol or other non-base.yml configs are at the end, but there should be more sophisticated ways to achieve this.

File: MaxText/tests/forward_pass_logit_checker.py (hunk 1 above)
Comment (RissyRan) [reply MaxText/tests/forward_pass_logit_checker.py:112]: Yeah, I agree. This may not be a good way. Let me change it back to all. But need to set loose limit. 

I am running some experiments [here](https://docs.google.com/spreadsheets/d/1o2RhkRpPXL5WD7FYb9kgm64qjuifOt0E0nfe3dc0YZE/edit?resourcekey=0-HGJ3wSqnzfMQ52O_Pf10sQ#gid=774077451), and absolute difference is quite large on TPU compared to CPU.

File: MaxText/layers/linears.py [hunk 4]
```diff
@@ -273,40 +274,81 @@ class MoeBlock(nn.Module):
...
+    w0_kernel = jnp.asarray(w0_kernel, self.dtype)
+    w1_kernel = self.param(
+        'wi_1',
+        nn.with_logical_partitioning(kernel_init, kernel_axes),
+        (num_experts, base_emb_dim, mlp_dim),
+        self.weight_dtype,
+        kernel_in_axis,
+        kernel_out_axis,
+      )
+    w1_kernel = jnp.asarray(w1_kernel, self.dtype)
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply MaxText/layers/linears.py:306]: Yeah, if I did not cast it to dtype, I will get incompatible type error. Similar to [this](https://github.com/google/maxtext/blob/61685809426db0e0d77a196e4f8ce29907cb5b81/MaxText/layers/linears.py#L134) in DenseGeneral.

File: MaxText/layers/linears.py [hunk 5]
```diff
@@ -273,40 +274,81 @@ class MoeBlock(nn.Module):
...
+            self.num_experts,
+            dtype=self.dtype,
+            kernel_init=self.kernel_init,
+            kernel_axes=self.kernel_axes,
+            name='gate')(inputs)
+    
+    top_k_weights, top_k_indices = jax.lax.top_k(gate_logits, self.num_experts_per_tok)
+    flattened_top_k_weights = top_k_weights.reshape(-1, self.num_experts_per_tok)
+
+    softmax_probs = jax.nn.softmax(flattened_top_k_weights.astype(jnp.float32), axis=-1).astype(self.weight_dtype)
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply MaxText/layers/linears.py:332]: Yeah, this matches implementation [here](https://github.com/mistralai/mistral-inference/blob/main/src/mistral_inference/moe.py#L29). Just in case, we changed the default `weight_dtype` in base.yml.

File: MaxText/tests/forward_pass_logit_checker.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (RissyRan) [reply MaxText/tests/forward_pass_logit_checker.py]: SG! updated.

File: MaxText/tests/forward_pass_logit_checker.py (hunk 1 above)
Comment (RissyRan) [reply MaxText/tests/forward_pass_logit_checker.py]: I will need to add a token size arg to Mixtral test, otherwise the atol and rtol need to be huge. It does not make sense to have a test with big value of those tolerance. Currently, Mixtral has 11 tokens due to bos and eos.

File: end_to_end/tpu/mixtral/8x7b/2_test_mixtral.sh [hunk 6]
```diff
@@ -30,6 +30,9 @@ export SCANNED_CHECKPOINT=${M_BASE_OUTPUT_DIRECTORY}/scanned_ckpt/0/items
 # Run decoding with converted ckpt
 python3 MaxText/decode.py MaxText/configs/base.yml load_parameters_path=${SCANNED_CHECKPOINT} run_name=scanned_decoding per_device_batch_size=1 model_name=mixtral-8x7b tokenizer_path=gs://maxtext-external/mixtral-8x7B-v0.1-Instruct/tokenizer.mistral ici_tensor_parallelism=4 ici_fsdp_parallelism=16 max_prefill_predict_length=11 max_target_length=24 prompt="[INST] I love to [/INST]" autoregressive_decode_assert="That's great to hear! I love to learn new things" attention=dot_product
 
+# Test whether the forward pass logits match the golden logits
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply end_to_end/tpu/mixtral/8x7b/2_test_mixtral.sh]: Updated.

File: MaxText/tests/forward_pass_logit_checker.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gobbleturk) [reply MaxText/tests/forward_pass_logit_checker.py]: The approach you have now is okay. An alternative would be to put rtol and atol first and only pass `sys.argv[2:]` to `pyconfig.initialize`

File: end_to_end/tpu/mixtral/8x7b/2_test_mixtral.sh [hunk 7]
```diff
@@ -26,10 +26,17 @@ export M_ASYNC_CHECKPOINTING=false
 
 # `SCANNED_CHECKPOINT` refers to the checkpoint that used for both `train.py` and `decode.py` 
 export SCANNED_CHECKPOINT=${M_BASE_OUTPUT_DIRECTORY}/scanned_ckpt/0/items
+export MATMUL_SCANNED_CHECKPOINT=${M_BASE_OUTPUT_DIRECTORY}/matmul_scanned_ckpt/0/items
 
 # Run decoding with converted ckpt
 python3 MaxText/decode.py MaxText/configs/base.yml load_parameters_path=${SCANNED_CHECKPOINT} run_name=scanned_decoding per_device_batch_size=1 model_name=mixtral-8x7b tokenizer_path=gs://maxtext-external/mixtral-8x7B-v0.1-Instruct/tokenizer.mistral ici_tensor_parallelism=4 ici_fsdp_parallelism=16 max_prefill_predict_length=11 max_target_length=24 prompt="[INST] I love to [/INST]" autoregressive_decode_assert="That's great to hear! I love to learn new things" attention=dot_product
 
+# Test whether the forward pass logits match the golden logits - for loop implementation
```
Comment (gobbleturk) [inline end_to_end/tpu/mixtral/8x7b/2_test_mixtral.sh:34]: Thank you for adding these!

**Review by gobbleturk — APPROVED** [review]

File: end_to_end/tpu/mixtral/8x7b/2_test_mixtral.sh [hunk 8]
```diff
@@ -26,10 +26,17 @@ export M_ASYNC_CHECKPOINTING=false
...
+python3 MaxText/tests/forward_pass_logit_checker.py  MaxText/configs/base.yml load_parameters_path=${SCANNED_CHECKPOINT} run_name=forward_pass_test per_device_batch_size=1 model_name=mixtral-8x7b tokenizer_path=gs://maxtext-external/mixtral-8x7B-v0.1-Instruct/tokenizer.mistral ici_tensor_parallelism=4 ici_fsdp_parallelism=16 max_prefill_predict_length=11 max_target_length=11 dataset_type=synthetic dtype=float32 --atol=1 --rtol=0.5 --token_size=4
+
+# Test whether the forward pass logits match the golden logits - matmul implementation
+python3 MaxText/tests/forward_pass_logit_checker.py  MaxText/configs/base.yml load_parameters_path=${MATMUL_SCANNED_CHECKPOINT} run_name=matmul_forward_pass_test per_device_batch_size=1 model_name=mixtral-8x7b tokenizer_path=gs://maxtext-external/mixtral-8x7B-v0.1-Instruct/tokenizer.mistral ici_tensor_parallelism=4 ici_fsdp_parallelism=16 max_prefill_predict_length=11 max_target_length=11 dataset_type=synthetic dtype=float32 --atol=3 --rtol=1 --token_size=4
```
Comment (A9isha) [inline end_to_end/tpu/mixtral/8x7b/2_test_mixtral.sh:38]: ah we are having to allow the high `atol=3`, let us track this to investigate further

**Review by A9isha — APPROVED** [review]
Thanks Ran!

File: end_to_end/tpu/mixtral/8x7b/2_test_mixtral.sh
[the remark this answers is not in the document]
Comment (RissyRan) [reply end_to_end/tpu/mixtral/8x7b/2_test_mixtral.sh:1]: ok, I will update that. Thanks!
